### ISSUE TITLE
Curiosity: Intrinsic reward doesn't follow forward loss

### ISSUE THREAD

**Issue by LeSphax** [Unity-Technologies/ml-agents#1374]
Hello,

I am training some agents to play a sport game similar to handball. For now I am training them to score goals with curriculum learning.

It looks like that [on video]( https://www.youtube.com/watch?v=IITLQJbAG8E) (When I teach them without curiosity)

The curriculum goes like this :
1 - Learn to pick up ball
2 - Learn to throw the ball and look toward the goals
3 - Learn to throw the ball in the goals
4 - Increase the size of the field

I activated curiosity in one of my recent experiments and the agent seems to learn well and score goals consistently until I increase the size of the field. Then he starts having trouble picking up the ball even though he still score goals when he gets it.

Here are the curiosity value.
![curiosity graph](https://user-images.githubusercontent.com/10918198/47480482-e235b300-d830-11e8-8c27-570989f7f41d.PNG)

So if I understand well the curiosity module is good at predicting the next state (forward loss) but not good at predicting the action that got it in that state(inverse loss).
The intrinsic reward is based on the forward loss and the inverse loss should only serve as a regularizer to get a good encoding.

But in the graphs, the intrinsic reward curve is quite similar to the inverse loss. At first, I thought the agent might intentionally output weird coordinates for where it tries to aim since it is impossible for him to figure out cartesian coordinates only from the state (rotation of the agent). But that shouldn't be possible since the inverse loss shouldn't affect the intrinsic reward.

So I don't understand is why the intrinsic reward is going up while the forward loss is staying constant. Shouldn't those two curves always have a similar shape?

**Comment by xiaomaogy** [comment]
@eshvk

**Comment by eshvk** [comment]
@LeSphax, the intrinsic reward is cumulative and for the entire episode, I am not sure why the two curves should have the similar shape. Have you seen [this](https://arxiv.org/abs/1705.05363) paper?

**Comment by LeSphax** [comment]
Hello eshvk,

Yes, I read that paper, I also read the code about curiosity for ml-agents but there are surely some parts I still don't understand otherwise I wouldn't be puzzled by this.
Looking at this code:
```
self.intrinsic_reward = tf.clip_by_value(self.curiosity_strength * squared_difference, 0, 1)
self.forward_loss = tf.reduce_mean(tf.dynamic_partition(squared_difference, self.mask, 2)[1])
```

The intrinsic rewards and the forward_loss are both based on the squared_difference so I assumed that they should change in the same direction. I thought that both forward_loss and the intrinsic rewards would increase/decrease when the squared_difference increases/decreases.

I hadn't considered that the intrinsic reward was cumulative while the forward_loss is an average of the values for the episode though. But if the episode length is constant it shouldn't matter, right? 

If SD is increasing, SD/K should be increasing as well as long as K is a constant. And looking back at tensorboard, the episode length was constant in my example.

Looking back at other runs that I did with curiosity activated, the curves are actually quite similar and the episode length explains differences well, for example here:

![image](https://user-images.githubusercontent.com/10918198/50139032-c8e53c00-02a0-11e9-8f00-9c3d4afd69df.png)
![image](https://user-images.githubusercontent.com/10918198/50139054-d7335800-02a0-11e9-91ed-db3b33087ec2.png)

Or here:

![image](https://user-images.githubusercontent.com/10918198/50139094-f205cc80-02a0-11e9-9ea7-26290601da88.png)

That's why the run I mentionned in this issue really surprised me with the intrinsic rewards increasing a lot while the forward loss staying constant.
